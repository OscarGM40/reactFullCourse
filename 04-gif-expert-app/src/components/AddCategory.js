import React, { useState } from "react";
import PropTypes from "prop-types";
import Swal from "sweetalert2";

const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState("");

  const handleInputChange = (e) => {
    setInputValue(e.target.value);
    //console.log("HandleInputChange llamado")
  };

  const handleSetCategory = (oldArray, category) => {
    if (!category) {
      Swal.fire({
        title: "Error",
        icon: "error",
        text: "Debes ingresar un texto",
        allowOutsideClick: false,
        allowEnterKey: false,
        allowEscapeKey: false,
      });

      setInputValue("");
      return [...oldArray];
    } else {
      const existe = oldArray.find(
        (categoryValue) => categoryValue === category
      );

      setInputValue("");
      return existe ? [...oldArray] : [category,...oldArray];
    }
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    //console.log("handlesubmit")

    /*   if(inputValue.trim().length > 2){
     setCategories((cat) => [...cat,inputValue])
     setInputValue('')
      } */
    setCategories((oldArray) => handleSetCategory(oldArray, inputValue));
    //console.log("Submit hecho");
  };
  return (
    <form onSubmit={handleSubmit}>
      {/* <p>{inputValue}</p> */}
      <input type="text" value={inputValue} onChange={handleInputChange} 
      placeholder="Buscar gifs..."/>
    </form>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};

export default AddCategory;
