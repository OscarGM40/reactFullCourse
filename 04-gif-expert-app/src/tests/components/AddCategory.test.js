import { shallow } from "enzyme";
import React from "react";
import AddCategory from "../../components/AddCategory";

describe("Pruebas en componente AddCategory", () => {
  const setCategories = jest.fn();

  let wrapper = shallow(<AddCategory setCategories={setCategories} />);

   beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategories={setCategories} />);
  }); 

  test("Debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de cambiar la caja de texto", () => {
    const input = wrapper.find("input");
    const value = "Hola mundo";
    input.simulate("change", {
      target: {
        value: value,
      },
    });
    const inputAfter = wrapper.find("input");
    expect(inputAfter.prop("value")).toBe(value);
  });

  test("No debe de hacer el onSubmit", () => {
     
    wrapper.find("form").simulate("submit", { preventDefault: () => {} });
    //como estamos simulando el submit, sin valor alguno, no deberia llamarse el setter del padre
    //sin embargo daba fallo porque la dearriba si puso un valor XD hay que resetear los Mocks↑↑
    //expect(setCategories).not.toHaveBeenCalled();
    expect(setCategories).toHaveBeenCalledTimes(1);
  });

   test('Debe de llamar el setCategories y limpiar la caja de texto', () => {
   //1- simular el inputChange
   //2- simular el submit
   //3- setCategories se debe de haber llamado
   //4- el valor del input debe de estar como un string vacio
      const input = wrapper.find("input");
      input.simulate("change",{target:{value:"mock"}});
      const inputAfter = wrapper.find("input");
      expect(inputAfter.prop("value")).toBe("mock");
      wrapper.find('form').simulate('submit',{preventDefault(){}});
      expect(setCategories).toHaveBeenCalledTimes(1);
      expect(setCategories).toHaveBeenCalledWith(expect.any(Function));
      expect(input.prop('value')).toBe("");





})


});
