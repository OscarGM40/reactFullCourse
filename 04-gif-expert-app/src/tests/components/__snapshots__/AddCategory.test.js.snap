// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente AddCategory Debe de mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    placeholder="Buscar gifs..."
    type="text"
    value=""
  />
</form>
`;
